### PR TITLE
tree shaking proof of concept

### PR DESCRIPTION
**Pull request by p9f**
Goal of this pull request is to make a few functions tree-shakable, as a proof of concept:
  - `replace`, `replaceAll`, `join`
  - `truncate`, `words`
 
If we agree on the implementation, we can then open more pull requests to make the rest of the library tree shakable (will be quite time consuming but not complex).

Tree shaking is mentioned in the [v2 roadmap discussion](https://github.com/gustavoguichard/string-ts/discussions/36#discussioncomment-7201877) - although I am not sure we need to wait for v2 since it should be possible do release it without any breaking change. As it implies a lot of diff, we might want to do it as soon as possible.

As a disclaimer, I don't know that much about modules and tree shaking. I use [Remeda](https://github.com/remeda/remeda) as an inspiration, since they put tree shaking in their [design goals](https://github.com/remeda/remeda#remeda-design-goals), and their source code was quite straight forward to read.

--

Implementation principles:
- dropping `tsup` and relying on raw tsc instead to build the package (I am unsure why we would need tsup?)
- having one file per tree shaking "entry point":
  - typically an entry point would be a single function. (for functions such as trim / trim start / trim end where the implementations are linked, we could have a single entry point for those).
  - utilities that need to be shared internally but not externally would have their filename starting with `_`. They also need to be split for the tree shaking.
  - `index.ts` would export `*` from all the files not starting with `_`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,23 +2,25 @@
...
+  "module": "dist/es/index.js",
+  "jsnext:main": "dist/es/index.js",
   "author": "Gustavo Guichard <@gugaguichard>",
   "license": "MIT",
   "scripts": {
-    "build": "tsup ./src/index.ts --format esm,cjs --dts",
+    "build": "npm run clean-dist && npm run build:es && npm run build:cjs",
     "dev": "tsup ./src/index.ts --format esm,cjs --watch --dts",
     "lint": "eslint *.ts*",
-    "tsc": "tsc --noEmit",
```
Comment (p9f) [inline package.json:14]: I moved this in the tsconfig.json - overwritten by the tsconfig.build.json when building.

File: package.json [hunk 2]
```diff
@@ -2,23 +2,25 @@
   "name": "string-ts",
   "version": "1.2.0",
   "description": "Strongly-typed string functions.",
-  "main": "./dist/index.js",
-  "module": "./dist/index.mjs",
-  "types": "./dist/index.d.ts",
+  "main": "dist/commonjs/index.js",
+  "module": "dist/es/index.js",
+  "jsnext:main": "dist/es/index.js",
```
Comment (gustavoguichard) [inline package.json:7]:
```suggestion
  "jsnext:main": "dist/es/index.js",
   "sideEffects": false,
```

Comment (gustavoguichard) [reply package.json:7]: I've read that adding this property also helps on tree-shaking as it indicates nothing in this package modifies global variables

**Comment by gustavoguichard** [comment]
Awesome work @p9f !
About having one file for function I guess it is ok, a good organization. Maybe I'd add type tests and runtime tests back into a single file in this case given the test files won't get too bloated.

My vision is to be able to do like in lodash:
```ts
import join from 'string-ts/join'
// or
import { join } from 'string-ts/join'
```

For that to work there must be other changes. Currently, to import it like that the path would be `string-ts/dist/es/join` though. Do you know what should be done to change that?

**Comment by p9f** [comment]
@gustavoguichard I don't know if we can do that and at the same time support both es module and common js.
If I am not mistaken, there is two packages in lodash (`lodash` for common js and `lodash-es` for es) - I don't think it's great to have two packages?

As I understand there is three way we might want to import `join`:
```typescript
// 1 - works with this PR
import * as stringTs from "string-ts";
stringTs.join(...)

// 2 - works with this PR
import { join } from "string-ts";

// 3 - does not work with this PR
import join from "string-ts/join"
```

1 & 2 would work with this PR - is there any reason you would like to have 3?

**Comment by p9f** [comment]
closing in favour in #52

**Comment by gustavoguichard** [comment]
@p9f and @jly36963 look at this tool which easily shows if a library is tree-shakeable:
https://pkg-size.dev/string-ts

**Pull request closed without merging**